File: all/234_palindrome_linked_list/sol_better.py
from typing import Optional


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
def reverseLinkedList(head: Optional[ListNode]) -> Optional[ListNode]:
    previous_node = None
    current_node = head
    next_node = None

    while current_node:
        next_node = (current_node.next)
        current_node.next = previous_node
        previous_node = current_node
        current_node = next_node

    return previous_node


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow_pointer = head
        fast_pointer = head

        while fast_pointer and (fast_pointer.next):
            slow_pointer = (slow_pointer.next)
            fast_pointer = ((fast_pointer.next).next)

        left_part_list_current_node = head
        right_part_list_current_node = reverseLinkedList(slow_pointer)

        while right_part_list_current_node:
            if (
                (left_part_list_current_node.val)
                !=
                (right_part_list_current_node.val)
            ):
                return False

            left_part_list_current_node = (left_part_list_current_node.next)
            right_part_list_current_node = (right_part_list_current_node.next)

        return True

File: all/234_palindrome_linked_list/sol_better.c
#include <stdbool.h>
#include <stddef.h>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseLinkedList(struct ListNode * const head) {
    struct ListNode *current_node = head;
    struct ListNode *previous_node = NULL;
    struct ListNode *next_node_ptr = NULL;

    while (NULL != current_node) {
        next_node_ptr = (current_node->next);
        current_node->next = previous_node;
        previous_node = current_node;
        current_node = next_node_ptr;
    }

    return previous_node;
}

bool isPalindrome(struct ListNode* head) {
    struct ListNode *left_part_list_current_node_ptr = NULL;
    struct ListNode *right_part_list_current_node_ptr = NULL;
    struct ListNode *slow_ptr = head;
    struct ListNode *fast_ptr = head;

    /* Find the middle node in the linked list. */
    while (
        (NULL != (fast_ptr->next))
        &&
        (NULL != ((fast_ptr->next)->next))
    ) {
        slow_ptr = (slow_ptr->next);
        fast_ptr = ((fast_ptr->next)->next);
    }

    /* Reverse the the linked list at the right hand side of the middle node. */
    right_part_list_current_node_ptr = reverseLinkedList(slow_ptr->next);
    left_part_list_current_node_ptr = head;

    while (NULL != right_part_list_current_node_ptr) {
        if (
            (left_part_list_current_node_ptr->val)
            !=
            (right_part_list_current_node_ptr->val)
        ) {
            return false;
        }

        left_part_list_current_node_ptr = (
            left_part_list_current_node_ptr->next
        );
        right_part_list_current_node_ptr = (
            right_part_list_current_node_ptr->next
        );
    }

    return true;
}
